### PR TITLE
Hotfix 0.9.61/274 rename upload project

### PR DESCRIPTION
**Pull request by DinosaurierRex**
CATROID-274 Fix for renaming projects in upload dialog

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java [hunk 1]
```diff
@@ -258,11 +259,13 @@ private void showProgressDialogAndUploadProject() {
...
-			ProjectLoadTask
-					.task(project.getDirectory(), this);
-			project = ProjectManager.getInstance().getCurrentProject();
+			try {
+				File renamedDirectory = ProjectRenameTask.task(project.getDirectory(), name);
+				ProjectLoadTask.task(renamedDirectory, this);
+				project = ProjectManager.getInstance().getCurrentProject();
+			} catch (IOException e) {
+				Log.e(TAG, "Creating renamed directory failed!", e);
+			}
```
Comment (dpedratscher) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: May be a new story and too much for a hotfix:
@DinosaurierRex can you add some info for the user, that the renaming failed? What does the user see, when it didn't work? 
@wslany  Is an info enough or should we give the chance to cancel the upload or retry?

Comment (thmq) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/activity/ProjectUploadActivity.java]: I don't really think it is necessary to think of some elaborate workflow here. This is a hotfix after all. We can think of some user interaction handling some other time.

File: catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectRenameTask.java [hunk 2]
```diff
@@ -54,21 +54,31 @@ public ProjectRenameTask setListener(ProjectRenameListener listener) {
...
 		File dstDir = new File(projectDir.getParent(), FileMetaDataExtractor.encodeSpecialCharsForFileSystem(dstName));
 
 		try {
-			return projectDir.renameTo(dstDir)
-					&& XstreamSerializer.renameProject(new File(dstDir, CODE_XML_FILE_NAME), dstName);
+			if (projectDir.renameTo(dstDir)
+					&& XstreamSerializer.renameProject(new File(dstDir, CODE_XML_FILE_NAME), dstName)) {
+				return dstDir;
+			} else {
+				return null;
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectRenameTask.java]: still not loving null returns and the null checks that are needed then. :crying_cat_face: 

how about throwing a different exception here? e.g. `throw new IOException("some cool message")`

File: catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectRenameTask.java [hunk 3]
```diff
@@ -54,21 +54,30 @@ public ProjectRenameTask setListener(ProjectRenameListener listener) {
...
+			if (projectDir.renameTo(dstDir)
+					&& XstreamSerializer.renameProject(new File(dstDir, CODE_XML_FILE_NAME), dstName)) {
+				return dstDir;
+			} else {
+				throw new IOException("Cannot rename project directory " + projectDir.getAbsolutePath() + " to " + dstName);
+			}
 		} catch (IOException e) {
-			Log.e(TAG, "Cannot rename project directory " + projectDir.getAbsolutePath() + " to " + dstName, e);
-			return false;
+			throw e;
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/io/asynctask/ProjectRenameTask.java]: please remove the `try { .. } catch ( .. )` block altogether now.

File: catroid/src/androidTest/java/org/catrobat/catroid/test/io/asynctask/ProjectRenameTaskTest.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	@Test
+	public void projectDirectoryRenameTest() throws IOException {
+		String absoluteRenamedDirectoryPath = defaultProject.getDirectory().getParent() + "/" + renamedProjectName;
+		File renamedDirectory = ProjectRenameTask.task(defaultProject.getDirectory(), renamedProjectName);
+		assertEquals(renamedDirectory.getAbsolutePath(), absoluteRenamedDirectoryPath);
+	}
+
+	@Test
+	public void projectDirectoryRenameSlashTest() throws IOException {
+		String absoluteRenamedDirectoryPath = defaultProject.getDirectory().getParent() + "/" + slashEncodedRenamedProject;
```
Comment (thmq) [inline catroid/src/androidTest/java/org/catrobat/catroid/test/io/asynctask/ProjectRenameTaskTest.java]: you could use:

`File expectedDirectory = new File(defaultProject.getDirectory().getParent(), slashEncodedRenamedProject);`

and then:

`assertEquals(expectedDirectory, renamedDirectory);`

**Comment by wslany** [comment]
Thanks!!

**Pull request merged**